feat(arc-17): sui custom token linking (#26)
axelarnetwork/arcs: ARCs/ARC-17.md

# ARC-17: Sui Custom Token Linking

## Metadata

- **ARC ID**: 17
- **Author(s)**: Drew Taylor
- **Category**: Interchain Token Protocol
- **Status**: Final
- **Created**: 2025-11-03
- **Last Updated**: 2025-11-04
- **Deployment Status**: Live

## Summary

This ARC specifies the implementation of custom token linking for Sui in the Interchain Token Service (ITS). Custom token linking enables existing Sui coins to be linked with tokens on other chains while supporting different decimal precisions through ITS Hub's automatic scaling. This specification details the Sui-specific patterns, `Channel`-based identity model, `CoinManagement` token manager structure, and key architectural differences from EVM implementations.

## Background and Motivation

ITS supports two distinct token linking approaches:

**Native Interchain Tokens**: Tokens deployed directly by ITS using standardized contracts. They must have identical decimals across all chains, and linking uses a **deploy origin** → **deploy remote** workflow. Metadata registration is not required for linking, since decimals are the same on both the origin and destination chains.

> **Custom Tokens**: Tokens deployed independently that integrate with ITS. Linking supports different decimal precisions across chains via ITS Hub's automatic scaling. Linking requires metadata registration on each chain, followed by registration in ITS, then explicit linking. Custom tokens cannot use the `NATIVE_INTERCHAIN_TOKEN` token manager type.

Reviewer: Also requires treasury cap be transferred to manager

Author: Only Mint / Burn managers require transferring the `TreasuryCap`, so I think it's better to cover this aspect under `CoinManagement`

### Motivation for Sui-Specific Token Linking ARC

Sui's implementation differs fundamentally from EVM (and other) chains in several areas:

1. **Channel-Based Identity**: Uses `Channel` objects instead of addresses for identity authorization
2. **Type Names for Token Addresses**: Sui coins are identified by their full type name, using the format `package::module::SYMBOL` (example: `0b6345a938be0a9bcd8319ca6b768bfddfffd60f0178d920973a528ef741c639::test::TEST`)
3. **CoinManagement Struct**: Replaces token manager contracts
4. **Object-Based Capabilities**: `TreasuryCap` objects replace role-based minter permissions
5. **Chain-Aware Token IDs**: Token ID derivation includes a chain name hash

### Goals

This ARC aims to:

1. Specify Sui's custom token linking implementation within the ITS framework

> 3. Clarify key differences from EVM token manager contracts

Reviewer: ```suggestion
2. Clarify key differences from EVM token manager contracts
```

Author: haha oops

> 4. Provide guidance for developers integrating Sui tokens with ITS

Reviewer: ```suggestion
3. Provide guidance for developers integrating Sui coins with ITS
```

6. **Message Format Compatibility**: Use ABI encoding compatible with EVM ITS implementations
7. **Salt Format**: Enforce 32-byte salt format matching Sui address format
8. **Salt Uniqueness**: Prevent reuse of same salt with same deployer `Channel`

### Security Requirements

1. **Channel Ownership Validation**: Verify `Channel` ownership during linking to ensure deployer authentication
2. **Token Manager Type Validation**: Prevent use of `NATIVE_INTERCHAIN_TOKEN` type for custom tokens
3. **Same-Chain Prevention**: Block linking tokens to the source chain itself
4. **Operator & Distributor Authorization**: Validate `Channel`-based operator and distributor permissions for administrative operations
5. **Treasury Cap Reclamation**: Enable original deployer to reclaim `TreasuryCap` if needed using `TreasuryCapReclaimer`
6. **Flow Limit Enforcement**: Support flow limits with operator-only modification rights

## Design

### Process Flow Overview

Sui custom token linking uses the following workflow:

```
1. User registers coin metadata on Sui via register_coin_metadata<T>

>    → Sends token type name and decimals to ITS Hub

Reviewer: ```suggestion
   → Sends token type, name, and decimals to ITS Hub
```

   
2. User registers coin metadata on destination chain
   → ITS Hub stores decimal mappings for both chains

3. User calls register_custom_coin<T> on Sui
   → Creates CoinManagement<T> (token manager)
   → Claims tokenId for deployer Channel + salt

>    → Returns TokenId and optional TreasuryCapReclaimer<T>

Reviewer: ```suggestion
   → Returns tokenId and optional TreasuryCapReclaimer<T>
```

Author: Hrmm, `TokenId` is the correct return type

4. User calls link_coin with the destination token address
   → Sends LINK_TOKEN message to ITS Hub
   → ITS Hub calculates scaling factor from stored decimals
   
5. Destination chain receives LINK_TOKEN message
   → Deploys corresponding token manager
   
6. ITS Hub now processes InterchainTransfer messages with automatic decimal scaling
```

### Channel-Based Identity Model

Sui uses `Channel` objects for authenticating rather than transaction sender addresses. A `Channel` is a Sui object with a unique UID that represents identity (much like an address) across ITS operations:

```move
public struct Channel has key, store {
    id: UID,
}
```

> Since `Channel` is an object the can be owned by any user or contract, transferring a `Channel` object to another party transfers all of its capabilities and permissions (including operatorship, and distributorship).

Reviewer: ```suggestion
Since a `Channel` is an object that can be owned by any user or contract, transferring a `Channel` object to another party transfers all of its capabilities and permissions (including operatorship, and distributorship).
```

> For convenience, `Channel`s can be converted to an address format using the `to_address` helper function:

Reviewer: ```suggestion
For convenience, `Channel`'s can be converted to an address format using the `to_address` helper function:
```

```move
public fun to_address(self: &Channel): address {
    object::id_address(self)
}
```

**Usage in Token Linking:**

> The same `Channel` must be used for both `register_custom_coin` and `link_coin` transactions. This proves the coin deployer's identity since the token ID is derived, in both cases, from the `Channel` and user provided `salt`, and the derived token ID must already exist (e.g. match the token ID from `register_custom_coin`) for token linking to succeed.

Reviewer: ```suggestion
The same `Channel` must be used for both `register_custom_coin` and `link_coin` transactions. This proves the coin deployer's identity, since the token ID is derived, in both cases, from the `Channel` and the  `salt`. The derived token ID must already exist (e.g. match the token ID from `register_custom_coin`) for the token linking to succeed.
```

### Token Address Encoding

#### Type Names vs Contract Addresses

Sui identifies coins by their **full type name** rather than by contract address. This fundamental difference stems from Move's type system, where each coin is defined as a generic type parameter `<T>` rather than a deployed contract instance.

**EVM Token Address Format:**
```
0x1234567890abcdef1234567890abcdef12345678  // 20-byte hex address
```

**Sui Token Type Name Format:**
```
0x6345a938be0a9bcd8319ca6b768bfddfffd60f0178d920973a528ef741c639::test::TEST // type name (package_id::module_name::TYPE)
```

The Sui format consists of three components separated by `::`:
1. **Package ID**: The hex address where the coin's package is deployed
2. **Module Name**: The Move module containing the coin definition
3. **Type Name**: The coin's type identifier (typically the coin symbol)

> This type-based identification is intrinsic to Move's design—coins aren't separate contract instances but rather instances of a generic `Coin<T>` type where `T` is the specific coin type.

Reviewer: ```suggestion
This type-based identification is intrinsic to Move's design, where coins aren't separate contract instances but rather instances of a generic `Coin<T>` type where `T` is the specific coin type.
```
